File: backend/src/hashing/ConsistentHashRing.js
/**
 * ConsistentHashRing.js
 *
 * WHY: Normal hash(key) % N routing means if a cache node is added/removed,
 * almost EVERY key changes node. That destroys the cache.
 *
 * Consistent Hashing solves this:
 * - Nodes are placed on a virtual ring (0 to 2^32)
 * - A key is hashed → placed on ring → walks CLOCKWISE to find its node
 * - If a node is removed, only keys between it and the previous node move
 *
 * This means only ~K/N keys are remapped (where K=keys, N=nodes).
 */

import crypto from 'crypto';

class ConsistentHashRing {
  /**
   * @param {string[]} nodes - Array of node names e.g. ['CacheNodeA','CacheNodeB','CacheNodeC']
   * @param {number} virtualNodes - Virtual replicas per physical node (more = better distribution)
   */
  constructor(nodes = [], virtualNodes = 150) {
    this.virtualNodes = virtualNodes;
    // ring: Map<hashPosition, nodeName>
    this.ring = new Map();
    // sorted array of hash positions for binary-search-style clockwise lookup
    this.sortedKeys = [];

    nodes.forEach(node => this.addNode(node));
  }

  /**
   * Hash a string key to a position on the ring (0 to 2^32 - 1)
   */
  _hash(key) {
    return parseInt(crypto.createHash('md5').update(key).digest('hex').slice(0, 8), 16);
  }

  /**
   * Add a physical node with its virtual replicas onto the ring
   */
  addNode(nodeName) {
    for (let i = 0; i < this.virtualNodes; i++) {
      const virtualKey = `${nodeName}#VN${i}`;
      const position = this._hash(virtualKey);
      this.ring.set(position, nodeName);
      this.sortedKeys.push(position);
    }
    this.sortedKeys.sort((a, b) => a - b);
  }

  /**
   * Remove a physical node (and all its virtual replicas) from the ring
   */
  removeNode(nodeName) {
    for (let i = 0; i < this.virtualNodes; i++) {
      const virtualKey = `${nodeName}#VN${i}`;
      const position = this._hash(virtualKey);
      this.ring.delete(position);
      const idx = this.sortedKeys.indexOf(position);
      if (idx !== -1) this.sortedKeys.splice(idx, 1);
    }
  }

  /**
   * Get the node responsible for a given key.
   * 1. Hash the key to get position
   * 2. Walk CLOCKWISE to find first node position >= key position
   * 3. If none found (key is past the last node), wrap around to first node
   */
  getNode(key) {
    if (this.sortedKeys.length === 0) {
      throw new Error('Hash ring is empty — no cache nodes registered');
    }

    const hashValue = this._hash(key);

    // Binary search: find first position >= hashValue
    let lo = 0, hi = this.sortedKeys.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (this.sortedKeys[mid] < hashValue) lo = mid + 1;
      else hi = mid;
    }

    // Wrap around if past the last position
    const position = this.sortedKeys[lo % this.sortedKeys.length];
    const nodeName = this.ring.get(position);

    return { nodeName, hashValue };
  }

  /**
   * Debug: show ring stats
   */
  getStats() {
    const distribution = {};
    for (const nodeName of this.ring.values()) {
      distribution[nodeName] = (distribution[nodeName] || 0) + 1;
    }
    return { totalVirtualNodes: this.sortedKeys.length, distribution };
  }
}

export default ConsistentHashRing;

File: backend-java/src/main/java/com/searchiq/hashing/ConsistentHashRing.java
package com.searchiq.hashing;

import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;

/**
 * ConsistentHashRing — distributes cache keys across multiple nodes.
 *
 * WHY consistent hashing (and not key % N):
 * - Normal modulo: adding/removing a node remaps ~all keys → cache wipe
 * - Consistent hashing: removing 1 of 3 nodes remaps only ~33% of keys
 *
 * HOW it works:
 * - Imagine a ring (clock face) from 0 to 2^32
 * - Each physical node gets VIRTUAL_NODES "virtual copies" placed on the ring
 * - A key is hashed → position on ring → walk clockwise → find first node
 * - If that node is removed, only keys between it and its predecessor move
 *
 * VIRTUAL NODES:
 * - Without them, 3 nodes on a ring of 2^32 positions would cluster unevenly
 * - 150 virtual nodes per physical node = even distribution (~33% each)
 */
@Component
public class ConsistentHashRing {

    private static final int VIRTUAL_NODES = 150;

    // Sorted map: ring position → node name (TreeMap keeps positions sorted)
    private final TreeMap<Long, String> ring = new TreeMap<>();

    // List of physical nodes currently on the ring
    private final List<String> nodeNames = new ArrayList<>();

    /**
     * Add a physical node with VIRTUAL_NODES replicas onto the ring.
     */
    public synchronized void addNode(String nodeName) {
        nodeNames.add(nodeName);
        for (int i = 0; i < VIRTUAL_NODES; i++) {
            long position = hash(nodeName + "#VN" + i);
            ring.put(position, nodeName);
        }
        System.out.println("[HASH RING] Added node: " + nodeName
                + " → total virtual nodes: " + ring.size());
    }

    /**
     * Remove a physical node and all its virtual replicas.
     */
    public synchronized void removeNode(String nodeName) {
        nodeNames.remove(nodeName);
        for (int i = 0; i < VIRTUAL_NODES; i++) {
            long position = hash(nodeName + "#VN" + i);
            ring.remove(position);
        }
    }

    /**
     * Get the node responsible for a key.
     *
     * Algorithm:
     * 1. Hash the key to a ring position
     * 2. Find the first entry in the TreeMap with key >= hashValue (clockwise)
     * 3. If none exists (hashValue > all positions), wrap around to ring.firstEntry()
     *
     * @param key the cache key (e.g. prefix "iph")
     * @return RingLookupResult with node name and hash value
     */
    public RingLookupResult getNode(String key) {
        if (ring.isEmpty()) {
            throw new IllegalStateException("Hash ring is empty — no nodes registered");
        }

        long hashValue = hash(key.toLowerCase());

        // Find first entry with position >= hashValue (clockwise walk)
        Map.Entry<Long, String> entry = ring.ceilingEntry(hashValue);

        // If none found (past the last entry), wrap around to first
        if (entry == null) {
            entry = ring.firstEntry();
        }

        return new RingLookupResult(entry.getValue(), hashValue);
    }

    /**
     * Returns distribution stats: how many virtual nodes each physical node has.
     */
    public Map<String, Integer> getDistribution() {
        Map<String, Integer> dist = new LinkedHashMap<>();
        for (String name : nodeNames) {
            dist.put(name, 0);
        }
        for (String nodeName : ring.values()) {
            dist.merge(nodeName, 1, Integer::sum);
        }
        return dist;
    }

    public int getTotalVirtualNodes() {
        return ring.size();
    }

    /**
     * MD5-based hash → unsigned 32-bit long.
     * Uses only the first 8 hex chars of the MD5 digest (same as Node.js implementation).
     */
    private long hash(String key) {
        try {
            MessageDigest md5 = MessageDigest.getInstance("MD5");
            byte[] digest = md5.digest(key.getBytes(StandardCharsets.UTF_8));
            // Take first 4 bytes and treat as unsigned 32-bit integer
            long result = 0;
            for (int i = 0; i < 4; i++) {
                result = (result << 8) | (digest[i] & 0xFF);
            }
            return result;
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("MD5 not available", e);
        }
    }

    /**
     * Result of a ring lookup.
     */
    public record RingLookupResult(String nodeName, long hashValue) {}
}
